### PR TITLE
Comment quality report for new files

### PR DESCRIPTION
**Pull request by AntoineAugusti**
Post a comment on the dataset page when a file is added or updated to report the quality against the schema.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -121,6 +136,36 @@ def build_details(details, report):
...
+
+    [Le fichier]({details["file_url"]}) que vous venez de publier ou mettre à jour comporte {plural(details["nb_errors"], "erreur")} sur un total de {plural(details["nb_rows"], "ligne")} par rapport au [schéma de référence]({schema_doc_url}).
+
+    Vous pouvez consulter le [dernier rapport de validation]({details["report_url"]}) pour vous aider à corriger les erreurs.
+    """
+
+    requests.post(
+        "https://www.data.gouv.fr/api/1/discussions/",
+        headers={
+            "X-API-KEY": os.environ["DATAGOUV_API_KEY"],
```
Comment (AntoineAugusti) [inline main.py:164]: @abulte Can you add the production key under Settings > Secrets for this repo?

File: main.py [hunk 2]
```diff
@@ -20,6 +24,17 @@ def schemas_details():
         return yaml.safe_load(response.content)
 
 
+@functools.lru_cache()
```
Comment (AntoineAugusti) [inline main.py:27]: See [functools](https://docs.python.org/3.3/library/functools.html)

> Decorator to wrap a function with a memoizing callable that saves up to the maxsize most recent calls.

Execute the function call only once

File: main.py [hunk 3]
```diff
@@ -20,6 +24,17 @@ def schemas_details():
...
+def existing_data():
+    with open(CSV_PATH, "r") as f:
+        return [d for d in csv.DictReader(f)]
+
+
+def file_is_new(file_url):
+    same_line = [row for row in existing_data() if file_url == row["file_url"]]
```
Comment (AntoineAugusti) [inline main.py:34]: `file_url` is the static.data.gouv.fr URL

File: main.py [hunk 4]
```diff
@@ -121,6 +136,36 @@ def build_details(details, report):
...
+
+    requests.post(
+        "https://www.data.gouv.fr/api/1/discussions/",
+        headers={
+            "X-API-KEY": os.environ["DATAGOUV_API_KEY"],
+            "User-Agent": "https://github.com/etalab/monitor-consolidation",
+        },
+        json={
+            "title": "Conformité au schéma",
+            "comment": textwrap.dedent(comment),
```
Comment (AntoineAugusti) [inline main.py:169]: Removes the beginning of the indentation for multiline strings

File: main.py (hunk 1 above)
Comment (abulte) [reply main.py:164]: Done, you've been added to the alias validation@ too.

File: main.py [hunk 5]
```diff
@@ -121,6 +136,36 @@ def build_details(details, report):
...
+    comment = f"""\
+    Bonjour,
+
+    [Le fichier]({details["file_url"]}) que vous venez de publier ou mettre à jour comporte {plural(details["nb_errors"], "erreur")} sur un total de {plural(details["nb_rows"], "ligne")} par rapport au [schéma de référence]({schema_doc_url}).
+
+    Vous pouvez consulter le [dernier rapport de validation]({details["report_url"]}) pour vous aider à corriger les erreurs.
+    """
+
+    requests.post(
+        "https://www.data.gouv.fr/api/1/discussions/",
```
Comment (abulte) [inline main.py:162]: can we test this on demo first 😇 maybe add a secret for the domain?

Comment (abulte) [reply main.py:162]: I added `DATAGOUV_API_DOMAIN` (= `demo.data.gouv.fr` for now)

Comment (abulte) [reply main.py:162]: but I guess you'll have to modify where you get the files from too 🤔 and this could become complicated.

Comment (AntoineAugusti) [reply main.py:162]: I committed this change. Is it doable or is it too much regarding file URLs?

**Comment by AntoineAugusti** [comment]
Is it okay to merge as is and run things on production? It's not easy to run this on demo because:
- dataset IDs are not the same
- files are uploaded or updated on production, not on demo
- `file_url`s are taken from production

I've executed the script locally to check the detection of new files and the API call.

I'd be happy to have your opinion on the message before shipping

**Comment by abulte** [comment]
@AntoineAugusti ok for prod.

About the message, this is what I wrote for the DECP validation: https://demo.data.gouv.fr/fr/datasets/fichier-csv/#discussion-5c9b8847634f4178a4eca78a

I think the part about "why this check has been triggered" is important.

File: main.py [hunk 6]
```diff
@@ -121,6 +136,36 @@ def build_details(details, report):
...
+        return f"{count} {word}"
+
+    schema_doc_url = f"https://schema.data.gouv.fr/{details['schema_slug']}/latest.html"
+
+    comment = f"""\
+    Bonjour,
+
+    [Le fichier]({details["file_url"]}) que vous venez de publier ou mettre à jour comporte {plural(details["nb_errors"], "erreur")} sur un total de {plural(details["nb_rows"], "ligne")} par rapport au [schéma de référence]({schema_doc_url}).
+
+    Vous pouvez consulter le [dernier rapport de validation]({details["report_url"]}) pour vous aider à corriger les erreurs.
```
Comment (geoffreyaldebert) [inline main.py:154]: Ce lien sera quel url ? 
Celui-ci : https://github.com/etalab/monitor-consolidation/blob/master/data/data.csv ?
Ou plus personnalisé ?

**Review by geoffreyaldebert — APPROVED** [review]
Très cool comme feature.
La question que je me pose : si un producteur fournit un jeu de données tous les jours et qu'il a toujours le même genre d'erreur, cela ne risque-t-il pas de l'éloigner des discussions datagouv ? Ne percevra-t-il pas ces multiples discussions comme un spam ?
C'est l'inconvénient d'intégrer ça dans les discussions en-dessous du jeu de données. Mais difficile de faire mieux pour l'instant. Je suis ok avec le test et on verra le ressenti des utilisateurs au fil de l'eau.

File: main.py (hunk 6 above)
Comment (AntoineAugusti) [reply main.py:154]: Le lien vers le rapport dans `report_url` pointe vers un rapport Validata, voici [un exemple](https://go.validata.fr/table-schema?input=url&schema_url=https://schema.data.gouv.fr/schemas/etalab/schema-lieux-covoiturage/0.1.2/schema.json&url=https://static.data.gouv.fr/resources/base-nationale-consolidee-des-lieux-de-covoiturage/20200428-190615/bnlc.csv&repair=true). Ce sont les liens que tu peux voir dans la colonne `report_url` dans le CSV

**Comment by AntoineAugusti** [comment]
@geoffreyaldebert Your points are all valid. This implementation is a quick and naive implementation to solve the current problem: producers want to know about their errors (they asked us) and we need to show that we're proactive at checking data quality.

That being said, this implementation has flaws:
- it can be seen as spammy. If you publish files often, you'll get multiple messages. Not even in the same thread. That could be improved later if it's an issue
- if the consolidation is done automatically, there is not much you can do to fix your errors
- some producers could be annoyed to have messages on their dataset saying that their data is not _clean_
- messages stick around on the dataset, even if the underlying problem is sometimes temporary

I think we're still figuring things out and we will learn from this feature in production.

**Comment by AntoineAugusti** [comment]
@abulte One last check before merging this, can you confirm that secrets have the right value for production on GH's settings? I can see that you've set them but I can't check their values.

File: main.py (hunk 6 above)
Comment (geoffreyaldebert) [reply main.py:154]: Nickel

**Comment by geoffreyaldebert** [comment]
I agree with your explanations. We'll see on the fly when it's in production.

**Comment by abulte** [comment]
@AntoineAugusti it should be ok now, but DOMAIN should be useless? I guess you can remove this env from your code.

**Pull request merged**